### ISSUE TITLE
Support named primary key constraints (e.g. with MSSQL)

### ISSUE THREAD

**Issue by madebysid** [prisma/prisma#4010]
## Problem

TL;DR: MSSQL uses named constraints for primary keys (composite and non-composite), but Prisma does not respect this name yet.

Suppose I have this MSSQL schema:

```
CREATE TABLE RegularPK (
	name varchar(32) NOT NULL,
	email varchar(32) NOT NULL,
	CONSTRAINT pk_const PRIMARY KEY (email)
);

CREATE TABLE CompositePK (
	name varchar(32) NOT NULL,
	email varchar(32) NOT NULL,
	CONSTRAINT composite_pk_const PRIMARY KEY (name, email)
);

CREATE TABLE CompositeUnique (
	name varchar(32) NOT NULL,
	email varchar(32) NOT NULL,
	CONSTRAINT composite_u_const UNIQUE (email, name)
);
```

This gets introspected as:

```
model RegularPK {
  email String @id
  name  String 
}

model CompositePK {
  name  String 
  email String

  @@id([name, key])
}

model CompositeUnique {
  name  String
  email String

  @@unique([name, email], name: "u_const")
}
```

This means that the generated typings make it so these are correct Prisma Client request:

```
prisma.compositePK.findOne({
  where: { email: 'x' },
})

prisma.compositePK.findOne({
  where: { name_email: { name: 'x', email: 'x' } },
})

prisma.compositeUnique.findOne({
  where: { u_const: { name: 'x', email: 'x' } },
})
```

Notable here is the fact that Prisma respects the name of the constraint in case of composite uniques & regular primary keys, but not with a composite primary key.

Moreover, the _only_ way to create a composite primary key in MSSQL as far as I know is to use a named constraint, so Prisma not respecting the name might be confusing for people, especially given the fact that we *do* respect the name of the constraint for composite uniques.
This isn't that big a deal for regular primary keys since they get mapped to a field (aka gets introspected correctly)

This is likely due to the fact that our `@@id` attribute does not support a `name` argument.

## Suggested solution

Support a `name` argument in the `@@id` attribute in our PSL.

How this would work with other databases is up for debate. 

Here are some things we might consider:
- Postgres, MySQL and SQLite do not support named composite primary keys
- Do we ignore the `name` argument in `@@id` for DBs that do not support it?
- Should the Client typings reflect the `name` argument even if the database does not support it?

For more context, if needed: https://prisma-company.slack.com/archives/C0110LHCCDB/p1603386503085600

**Comment by matthewmueller** [comment]
### Proposed Solution

@mavilein and I just chatted about this one. We concluded that another way to reduce confusion is to unlink the constraint name with what gets generated in the client. This would be similar to what we do with `@map`:

```prisma
model User {
  name String @map("my_complex_database_field")
}
```

This feels like a better solution because it allows you to have more control with the database client that gets generated. It would also allow us to defer introducing a `@@id(name)` field which only seems popular on MSSQL, though probably possible on other databases (untested):

```sql
create table foo (
  bar int,
  constraint bar_pk primary key (bar)
)
```

### Alternatives

#### 1. `@@unique(name:"nameAndEmail", map:"u_const")`

`name` adjusts to the generated client, `map` maps to the underlying constraint name. This would be a breaking change because we've changed the meaning of `name`. Wouldn't be hard to write a script to help people upgrade.

#### 2. `@@unique(name:"u_const", alias:"nameAndEmail")`

Keep `name` as is it. Add `alias` to control the generated field name. Probably still a breaking change, because `@@unique(name:"u_const")` alone should not affect the generated client, which it currently does.

---

Both options are viable. @mavilein and I don't have strong opinions either way. We're curious to hear what you think!

**Comment by matthewmueller** [comment]
This solution would also solve #3193

**Comment by do4gr** [comment]
This is now available via the `namedConstraints` preview flag since 2.29. See https://www.prisma.io/docs/guides/upgrade-guides/upgrading-to-use-preview-features/enabling-named-constraints
